### PR TITLE
Update the `updater.rs` code.

### PR DESCRIPTION
**Pull request by MathieuDutSik**
## Motivation

The `updater.rs` is currently using the `blob_last_used_by` with the error case returning the missing entries which is a bd design:
* It might get you a wrong index if the error is not `not_found`.
* It does not group the data.

## Proposal

The following was done:
* Introduce the `missing_blob_states` that provides what we are after.
* Remove the `blobs_last_used_by` that was previously introduced, but keep it in the `db_storage.rs`.
* The replacement of stream and the attention to only getting whether a blob_state is present or not allows for grouping the operation efficiently.

Possible criticism of the design:
* We could have kept the `blobs_last_used_by` but there was no use case and in particular, we could not use it for getting the missed blob ids.
* The API is left in the `db_storage.rs`. Possibly we could remove it as well as other unused APIs.
* The `missing_blob_ids` does a single operation. We could split it into blocks and then use a stream in a future PR. However, the size is quite small, only `BlobId`s.

There is no remaining `read_certificate(s)` that is done for extracting only small data.

## Test Plan

The CI

## Release Plan

This changes the binaries but does not change the storage. So, a switch 

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-core/src/updater.rs [hunk 1]
```diff
@@ -282,14 +282,9 @@ where
...
                     // certificates that last used the blobs to the validator node should be enough.
-                    let missing_blob_ids = stream::iter(mem::take(&mut blob_ids))
-                        .filter(|blob_id| {
-                            let node = self.remote_node.node.clone();
-                            let blob_id = *blob_id;
-                            async move { node.blob_last_used_by(blob_id).await.is_err() }
-                        })
-                        .collect::<Vec<_>>()
-                        .await;
+                    let blob_ids = mem::take(&mut blob_ids).into_iter().collect::<Vec<_>>();
```
Comment (afck) [inline linera-core/src/updater.rs:285]: Saving one allocation:

```suggestion
                    let blob_ids = blob_ids.drain().collect::<Vec<_>>();
```

**Review by afck — APPROVED** [review]
Good catch! We actually are only interested in the missing _blobs_ in this case.

File: linera-storage/src/db_storage.rs [hunk 2]
```diff
@@ -390,6 +400,24 @@ where
         Ok(missing_blobs)
     }
```
Comment (ndr-ds) [inline linera-storage/src/db_storage.rs:401]: <i class='graphite__hidden'>[Re: lines +385 to +401]<br/><br/></i>

Like @afck mentioned, `missing_blob_states` should be semantically equivalent to this one (we'll only have a state in storage if we have a blob also). So we can probably just use this existing one instead and delete `missing_blob_states`<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/2894?utm_source=unchanged-line-comment">Graphite</a>.</span>

**Review by ndr-ds — APPROVED** [review]

File: linera-core/src/updater.rs (hunk 1 above)
Comment (MathieuDutSik) [reply linera-core/src/updater.rs:285]: Thank you.

File: linera-core/src/node.rs [hunk 3]
```diff
@@ -105,9 +105,8 @@ pub trait ValidatorNode {
     /// Returns the hash of the `Certificate` that last used a blob.
     async fn blob_last_used_by(&self, blob_id: BlobId) -> Result<CryptoHash, NodeError>;
 
-    /// Returns the hashes of the `Certificate`s that last used some blobs.
-    async fn blobs_last_used_by(&self, blob_ids: Vec<BlobId>)
-        -> Result<Vec<CryptoHash>, NodeError>;
+    /// Returns the missing `BlobState`s.
+    async fn missing_blob_states(&self, blob_ids: Vec<BlobId>) -> Result<Vec<BlobId>, NodeError>;
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-core/src/node.rs:109]: Ok, thank you. I was not sure about that and did not want to take chances.

File: linera-storage/src/db_storage.rs (hunk 2 above)
Comment (MathieuDutSik) [reply linera-storage/src/db_storage.rs:401]: Ok, thank you.

**Pull request merged**